### PR TITLE
Feature/add date separators option for empty threads

### LINKED ISSUE #4619: [V5] Add config flags to enable date separator items in case of empty threads (between parent and thread placeholder)
### Context & proposal

In our MessageListItemLiveData we should add a way to add date separator items even if there are no items in the thread. This is useful for people using the ThreadPlaceholder items.

### Acceptance criteria

To accept this task we need to complete the following items:

- [ ] Expose a config flag in ChatUI or something similar to enable this behavior.
- [ ] Add the date separator item to to the list of data based on the flag.

### PR DESCRIPTION
**Pull request by filbabic**
### 🎯 Goal

Closes #4619 

### 🛠 Implementation details

Added a boolean. It changes how the live data adds items. If the thread is empty and the boolean is true, adds a separator based on parent.

### 🎨 UI Changes

| Before | After |
| --- | --- |
| Screenshot_1674021621 | Screenshot_1674020670 |

Note that the before image doesn't have a custom thread item, but the after image does.

### 🧪 Testing

1. Apply patch
2. See if the item appears when empty thread is opened
3. See if the item/data is structured correctly if a message is sent to the thread
4. Run v5 and check that this behavior doesn't exist

<details>

<summary>Provide the patch summary here</summary>

```
Index: stream-chat-android-ui-components/src/main/kotlin/io/getstream/chat/android/ui/message/list/adapter/MessageListItemViewHolderFactory.kt
IDEA additional info:
Subsystem: com.intellij.openapi.diff.impl.patch.CharsetEP
<+>UTF-8
===================================================================
diff --git a/stream-chat-android-ui-components/src/main/kotlin/io/getstream/chat/android/ui/message/list/adapter/MessageListItemViewHolderFactory.kt b/stream-chat-android-ui-components/src/main/kotlin/io/getstream/chat/android/ui/message/list/adapter/MessageListItemViewHolderFactory.kt
--- a/stream-chat-android-ui-components/src/main/kotlin/io/getstream/chat/android/ui/message/list/adapter/MessageListItemViewHolderFactory.kt	(revision df97c24f938b16261435d5cbd827581e736a92fc)
+++ b/stream-chat-android-ui-components/src/main/kotlin/io/getstream/chat/android/ui/message/list/adapter/MessageListItemViewHolderFactory.kt	(date 1674021012562)
@@ -20,6 +20,8 @@
 import android.view.ViewGroup
 import com.getstream.sdk.chat.adapter.MessageListItem
 import io.getstream.chat.android.ui.ChatUI
+import io.getstream.chat.android.ui.common.extensions.internal.streamThemeInflater
+import io.getstream.chat.android.ui.databinding.StreamUiCustomTestBinding
 import io.getstream.chat.android.ui.message.list.GiphyViewHolderStyle
 import io.getstream.chat.android.ui.message.list.MessageListItemStyle
 import io.getstream.chat.android.ui.message.list.MessageReplyStyle
@@ -38,6 +40,7 @@
 import io.getstream.chat.android.ui.message.list.adapter.MessageListItemViewType.THREAD_PLACEHOLDER
 import io.getstream.chat.android.ui.message.list.adapter.MessageListItemViewType.THREAD_SEPARATOR
 import io.getstream.chat.android.ui.message.list.adapter.MessageListItemViewType.TYPING_INDICATOR
+import io.getstream.chat.android.ui.message.list.adapter.internal.DecoratedBaseMessageItemViewHolder
 import io.getstream.chat.android.ui.message.list.adapter.internal.MessageListItemViewTypeMapper
 import io.getstream.chat.android.ui.message.list.adapter.viewholder.attachment.AttachmentFactoryManager
 import io.getstream.chat.android.ui.message.list.adapter.viewholder.decorator.internal.DecoratorProvider
@@ -163,7 +166,7 @@
             GIPHY -> createGiphyMessageItemViewHolder(parentView)
             SYSTEM_MESSAGE -> createSystemMessageItemViewHolder(parentView)
             ERROR_MESSAGE -> createErrorMessageItemViewHolder(parentView)
-            THREAD_PLACEHOLDER -> createEmptyMessageItemViewHolder(parentView)
+            THREAD_PLACEHOLDER -> createCustomView(parentView)
             LINK_ATTACHMENTS -> createLinkAttachmentsViewHolder(parentView)
             GIPHY_ATTACHMENT -> createGiphyAttachmentViewHolder(parentView)
             FILE_ATTACHMENTS -> createFileAttachmentsViewHolder(parentView)
@@ -172,6 +175,20 @@
         }
     }
 
+    private fun createCustomView(parentView: ViewGroup): BaseMessageItemViewHolder<out MessageListItem> {
+        val binding = StreamUiCustomTestBinding.inflate(
+            parentView.streamThemeInflater,
+            parentView,
+            false
+        )
+
+        return object : DecoratedBaseMessageItemViewHolder<MessageListItem.ThreadPlaceholderItem>(binding.root, decoratorProvider.decorators) {
+            override fun bindData(data: MessageListItem.ThreadPlaceholderItem, diff: MessageListItemPayloadDiff?) {
+                super.bindData(data, diff)
+            }
+        }
+    }
+
     /**
      * Creates the custom attachments view holder.
      *
Index: stream-chat-android-ui-components-sample/src/main/kotlin/io/getstream/chat/ui/sample/application/ChatInitializer.kt
IDEA additional info:
Subsystem: com.intellij.openapi.diff.impl.patch.CharsetEP
<+>UTF-8
===================================================================
diff --git a/stream-chat-android-ui-components-sample/src/main/kotlin/io/getstream/chat/ui/sample/application/ChatInitializer.kt b/stream-chat-android-ui-components-sample/src/main/kotlin/io/getstream/chat/ui/sample/application/ChatInitializer.kt
--- a/stream-chat-android-ui-components-sample/src/main/kotlin/io/getstream/chat/ui/sample/application/ChatInitializer.kt	(revision df97c24f938b16261435d5cbd827581e736a92fc)
+++ b/stream-chat-android-ui-components-sample/src/main/kotlin/io/getstream/chat/ui/sample/application/ChatInitializer.kt	(date 1674020348014)
@@ -81,5 +81,6 @@
 
         // Using markdown as text transformer
         ChatUI.messageTextTransformer = MarkdownTextTransformer(context)
+        ChatUI.showDateSeparatorInEmptyThread = true
     }
 }
Index: stream-chat-android-ui-components/src/main/res/layout/stream_ui_custom_test.xml
IDEA additional info:
Subsystem: com.intellij.openapi.diff.impl.patch.CharsetEP
<+>UTF-8
===================================================================
diff --git a/stream-chat-android-ui-components/src/main/res/layout/stream_ui_custom_test.xml b/stream-chat-android-ui-components/src/main/res/layout/stream_ui_custom_test.xml
new file mode 100644
--- /dev/null	(date 1674021014747)
+++ b/stream-chat-android-ui-components/src/main/res/layout/stream_ui_custom_test.xml	(date 1674021014747)
@@ -0,0 +1,34 @@
+<?xml version="1.0" encoding="utf-8"?>
+<!--
+    Copyright (c) 2014-2023 Stream.io Inc. All rights reserved.
+
+    Licensed under the Stream License;
+    you may not use this file except in compliance with the License.
+    You may obtain a copy of the License at
+
+      https://github.com/GetStream/stream-chat-android/blob/main/LICENSE
+
+    Unless required by applicable law or agreed to in writing, software
+    distributed under the License is distributed on an "AS IS" BASIS,
+    WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+    See the License for the specific language governing permissions and
+    limitations under the License.
+-->
+<androidx.constraintlayout.widget.ConstraintLayout xmlns:android="http://schemas.android.com/apk/res/android"
+    xmlns:app="http://schemas.android.com/apk/res-auto"
+    android:layout_width="match_parent"
+    android:layout_height="wrap_content"
+    >
+
+    <TextView
+        android:id="@+id/customText"
+        android:layout_width="match_parent"
+        android:layout_height="300dp"
+        android:gravity="bottom"
+        android:text="Custom placeholder item test"
+        app:layout_constraintEnd_toEndOf="parent"
+        app:layout_constraintStart_toStartOf="parent"
+        app:layout_constraintTop_toTopOf="parent"
+        />
+
+</androidx.constraintlayout.widget.ConstraintLayout>
\ No newline at end of file

```

</details>

#### General
- [x] Assigned a person / code owner group (required)
- [x] Thread with the PR link started in a respective Slack channel (#android-chat-core or #android-chat-ui) (required)
- [x] PR is linked to the GitHub issue it resolves

#### Code & documentation
- [ ] Changelog is updated with client-facing changes
- [ ] Comparison screenshots added for visual changes
- [ ] Affected documentation updated (KDocs, docusaurus, tutorial)

### ☑️Reviewer Checklist
- [ ] UI Components sample runs & works
- [ ] UI Changes correct (before & after images)
- [ ] New feature tested and works
- [ ] Release notes and docs clearly describe changes
- [ ] All code we touched has new or updated KDocs

### CODE REVIEW

File: stream-chat-android-ui-components/api/stream-chat-android-ui-components.api
Comment (MarinTolic) [inline stream-chat-android-ui-components/api/stream-chat-android-ui-components.api:656]: We should write a changelog message  and add this to the [configuration docs](https://getstream.io/chat/docs/sdk/android/ui/general-customization/chatui/) before merging.

**Review by MarinTolic — COMMENTED** [review]
Works great but we should write a changelog message and add the change to the [configuration docs](https://getstream.io/chat/docs/sdk/android/ui/general-customization/chatui/)  before merging.

### ☑️Reviewer Checklist
- [x] UI Components sample runs & works
- [x] UI Changes correct (before & after images)
- [x] New feature tested and works
- [x] Release notes and docs clearly describe changes
- [ ] All code we touched has new or updated KDocs

**Review by bychkovdmitry — APPROVED** [review]
Works as described!

### ☑️Reviewer Checklist
- [x] UI Components sample runs & works
- [x] UI Changes correct (before & after images)
- [x] New feature tested and works
- [ ] Release notes and docs clearly describe changes
- [x] All code we touched has new or updated KDocs

File: CHANGELOG.md
Comment (MarinTolic) [inline CHANGELOG.md:65]:
```suggestion
- Added a feature flag to `ChatUI` called `showDateSeparatorInEmptyThread`. You can use this to enable a date separator if the thread is empty. [#4629](https://github.com/GetStream/stream-chat-android/pull/4629)
```

File: docusaurus/docs/Android/03-ui/02-general-customization/02-chatui.mdx
Comment (MarinTolic) [inline docusaurus/docs/Android/03-ui/02-general-customization/02-chatui.mdx:30]: If we're thinking about `DateSeparatorItem` maybe it's best to put it down as such or alternatively just "date separator item"

```suggestion
* `showDateSeparatorInEmptyThread`: Allows showing the DateSeparator item in the `MessageListView`, when opening an empty thread. This helps implement more complex use cases for empty threads and `ThreadPlaceholderItem`s.
```

File: docusaurus/docs/Android/03-ui/02-general-customization/02-chatui.mdx
Comment (MarinTolic) [inline docusaurus/docs/Android/03-ui/02-general-customization/02-chatui.mdx:524]: Same as the comment above concerning `DateSeparatorItem`

**Review by MarinTolic — COMMENTED** [review]
Everything looks good but I've been experiencing Docusaurus issues (not related to the PR) meaning I can't check the page locally hence comment without approval. Will re-check tomorrow.

**Review by TGazica — APPROVED** [review]
### ☑️Reviewer Checklist
- [x] UI Components sample runs & works
- [x] UI Changes correct (before & after images)
- [x] New feature tested and works
- [x] Release notes and docs clearly describe changes
- [x] All code we touched has new or updated KDocs

**Pull request merged**
